### PR TITLE
[REVIEW] CMake: use CPM for thirdparty dependencies

### PR DESCRIPTION
**Pull request by germasch**
This adds [CPM](https://github.com/TheLartians/CPM.cmake) to pull in thirdparty dependencies (see rapidsai/cudf#6095). It pulls in the current release of `CPM.cmake`, so it makes sure that RMM remains functional even if the CPM repo were to move or go away. For now, this PR doesn't change behavior, it just acts as a wrapper replacing the direct `FetchContent` calls.

As such, it just shortens things a bit, but down the road it also simplifies using pre-existing dependencies (e.g., when used with a package manager) rather than having them all be downloaded and compiled as part of RMM. Which then eventually will improve integrating RMM into package managers as well, and using it in downstream cmake projects.

A minor additional change is that this PR upgrades googletest from 1.8.0 -> 1.8.1 to avoid some cmake warnings in googletest that were fixed in that minor release.

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by harrism** [comment]
ok to test

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by harrism** [comment]
Should we consider updating to the latest gtest? (CC @kkraus14 not sure if we are limited to 1.8.x for any reason)

File: cmake/CPM.cmake
Comment (harrism) [inline cmake/CPM.cmake:1]: Is it not possible to get this file via FetchContent so we don't have to vendor it into RMM directly?

File: thirdparty/CMakeLists.txt
Comment (harrism) [inline thirdparty/CMakeLists.txt:6]: Does CPM have the equivalent of GIT_SHALLOW (doesn't fetch the entire git history, which can be quite large)?

File: cmake/CPM.cmake
Comment (germasch) [reply cmake/CPM.cmake:1]: CPM actually does recommend  to use this as `CPM.cmake`: https://github.com/TheLartians/CPM.cmake/blob/master/cmake/get_cpm.cmake, which will automatically download CPM. 
I instead copied in the whole file because of the comments [here](https://github.com/rapidsai/cudf/pull/6095#discussion_r477643040), but personally I don't mind either way.

File: thirdparty/CMakeLists.txt
Comment (germasch) [reply thirdparty/CMakeLists.txt:6]: Yes, I believe I can add back `GIT_SHALLOW` and it'll work as intended. CPM does automatically use shallow under certain circumstances:

| * Automatic shallow clone: if a version tag (e.g. v2.2.0) is provided and CPM_SOURCE_CACHE is used, CPM.cmake will perform a shallow clone of the dependency, which should be significantly faster while using less storage than a full clone.

But since `CPM_SOURCE_CACHE` isn't set (though it could be) by default this won't apply. I'll be happy to add back `GIT_SHALLOW` if that's preferred (I agree that it helps keeping existing behavior).

I'm not quite sure what motivates CPMs behavior. My guess is that it's to support offline work -- ie., if one sets `CPM_SOURCE_CACHE`, every dependency (say, in all of rapids) will be cloned (fully) only once, and then will stick around, and when it's used in a project like rmm or cudf, it'll just do a shallow clone into that project.

**Comment by germasch** [comment]
> Should we consider updating to the latest gtest? (CC @kkraus14 not sure if we are limited to 1.8.x for any reason)

FWIW, I just tried with gtest 1.10.0 and it compiled just fine.

File: cmake/CPM.cmake
Comment (harrism) [reply cmake/CPM.cmake:1]: I think they are suggesting that we create a RAPIDS fork of the file, although I think as long as we have a way to get a specific tag/version that won't change underneath us we should be OK. I definitely prefer either forking (which one of the RAPIDS org would need to do) or fetching a specific version to vendoring the file into RMM.   We are able to fetch specific versions of other dependencies without vendoring, I don't see why CPM whould be any different.

**Comment by kkraus14** [comment]
No reason we shouldn't upgrade from my perspective. If I remember correctly I tried bumping libcudf to 1.10.0 as well and ran into issues.

File: cmake/CPM.cmake
Comment (germasch) [reply cmake/CPM.cmake]: I updated the PR to use their `get_cpm.cmake` scripts used as `CPM.cmake`, which will automatically download the
real `CPM.cmake` (from a specified release) and include it. That also supports caching on the local machine, ie, things can work in offline mode if `CPM_SOURCE_CACHE` is used.

**Comment by germasch** [comment]
> No reason we shouldn't upgrade from my perspective. If I remember correctly I tried bumping libcudf to 1.10.0 as well and ran into issues.

I can push a change here that moves gtest to 1.10.0, or I can do it in a separate follow-up PR (I'd prefer the latter, since it's an unrelated change..)

File: cmake/CPM.cmake
Comment (kkraus14) [inline cmake/CPM.cmake:14]: I know this is the prescribed way of doing this in CPM, but using a git tag instead of commit hash means that someone could in theory change it out from underneath us. We should use a commit hash instead:

Here's the link to the commit hash at 0.27.3: https://raw.githubusercontent.com/TheLartians/CPM.cmake/c3c7e2d9a3b890511891ffc415d7ea23c689068a/cmake/CPM.cmake

File: thirdparty/CMakeLists.txt
Comment (kkraus14) [inline thirdparty/CMakeLists.txt:23]: I'm guessing this is to just keep the current flow, but we should move to using a Thrust target in the future to allow properly exporting a target for rmm.

**Review by kkraus14 — CHANGES_REQUESTED** [review]

File: thirdparty/CMakeLists.txt
Comment (germasch) [reply thirdparty/CMakeLists.txt:23]: Yup, it's in the plans (but probably saved for last, since it'll be the most unpleasant one to deal with).

**Comment by kkraus14** [comment]
> I can push a change here that moves gtest to 1.10.0, or I can do it in a separate follow-up PR (I'd prefer the latter, since it's an unrelated change..)

Agreed a follow up would be better

File: cmake/CPM.cmake
Comment (germasch) [reply cmake/CPM.cmake]: I updated the script to pull by hash. The script it pulls actually has `set(CURRENT_CPM_VERSION 0.27.2-development-version)`, but it's otherwise identical to 0.27.3's. I think that's because they change the version info only within their actual release process.

**Comment by germasch** [comment]
> > I can push a change here that moves gtest to 1.10.0, or I can do it in a separate follow-up PR (I'd prefer the latter, since it's an unrelated change..)
> 
> Agreed a follow up would be better

Basically already happened ;) #568

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
> Basically already happened ;) #568

Just merged, so feel free to bump it in this PR for CPM 😄

**Comment by germasch** [comment]
> Just merged, so feel free to bump it in this PR for CPM 😄

Done (had to, anyway, to resolve resulting conflict). 

I also added back `GIT_SHALLOW`, to keep that part of existing behavior the same for now.

**Comment by germasch** [comment]
What's the recommended way to retrigger the CI? (usually I force-push)

File: thirdparty/CMakeLists.txt
Comment (jrhemstad) [inline thirdparty/CMakeLists.txt:56]: When I [did this in cuCollections](https://github.com/NVIDIA/cuCollections/blob/035ed1bdabfbe00f662672c16cb0fd5aca9578a2/benchmarks/CMakeLists.txt#L5-L15) I had to workaround for the fact that this was broken with gcc5.4. I guess we dropped support for 5.4 in RAPIDS, so maybe it doesn't matter.

**Comment by jrhemstad** [comment]
rerun tests

File: thirdparty/CMakeLists.txt
Comment (germasch) [reply thirdparty/CMakeLists.txt:56]: Thanks for letting me know. So far it's been working w/o trouble, but it'll be good to know if someone does hit it at some point (though with RMM having dropped support for older CUDA toolkits, hopefully not).

**Review by kkraus14 — APPROVED** [review]

File: thirdparty/CMakeLists.txt
Comment (jrhemstad) [inline thirdparty/CMakeLists.txt:23]: We were previously using the hash because there wasn't a release with all the fixes we needed yet. 1.10.0 was just released that has everything we need. 
```suggestion
  VERSION 1.10.0
```

**Review by jrhemstad — APPROVED** [review]

File: cmake/CPM.cmake
Comment (harrism) [reply cmake/CPM.cmake]: Why would they change a release tag in a way that would break users?

**Comment by harrism** [comment]
@germasch is this good to merge? You made a comment above about saving something for last so I'm not sure.

**Comment by kkraus14** [comment]
> @germasch is this good to merge? You made a comment above about saving something for last so I'm not sure.

I think Jake's comment needs to be addressed first. The comment about saving for last was moving the CMake to using a Thrust target instead of pointing to directories which should be done in a follow up.

**Comment by germasch** [comment]
I updated thrust to 1.10.0. As far as I'm concerned, it's ready to merge once it passes CI. 

About using a target for thrust, I'm planning to do it in a future PR (It's not hard to use the fetched thrust this way, but it'll be more interesting if/when one wants to give the user a choice of using their own thrust install, or the one that came with CUDA).

File: thirdparty/CMakeLists.txt
Comment (kkraus14) [inline thirdparty/CMakeLists.txt:23]: Should this be marked `GIT_SHALLOW TRUE` as well?

**Review by harrism — DISMISSED** [review]
Since we are adding a new dependency on a third party library we have to jump through some process hoops internally at NVIDIA first. Shouldn't be long. Just blocking merge until then.

File: thirdparty/CMakeLists.txt
Comment (germasch) [reply thirdparty/CMakeLists.txt:22]: Yes, good point. Now fixed.

**Review by kkraus14 — APPROVED** [review]

**Comment by germasch** [comment]
rerun tests

**Comment by germasch** [comment]
> rerun tests

It was worth a shot, but I don't think the system listens to me ;)

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
> Since we are adding a new dependency on a third party library we have to jump through some process hoops internally at NVIDIA first. Shouldn't be long. Just blocking merge until then.

Just got sign off from internal NVIDIA process, so dismissing Mark's review and merging!

**Pull request merged**
